conans/client/build/meson.py

```diff
@@ -159,15 +162,22 @@ def configure(self, args=None, defs=None, source_dir=None, build_dir=None,
 
     @property
     def _vcvars_needed(self):
-        return self._compiler == "Visual Studio" and self.backend == "ninja"
+        return (self._compiler == "Visual Studio" and self.backend == "ninja" and
+                platform.system() == "Windows")
 
     def _run(self, command):
-        with tools.vcvars(self._settings,
-                          output=self._conanfile.output) if self._vcvars_needed else tools.no_op():
+        def _build():
             env_build = AutoToolsBuildEnvironment(self._conanfile)
             with environment_append(env_build.vars):
                 self._conanfile.run(command)
 
+        if self._vcvars_needed:
+            vcvars_dict = tools.vcvars_dict(self._settings, output=self._conanfile.output)
+            with environment_append(vcvars_dict, post=self._append_vcvars):
```

Comment:
Maybe the `post=self._append_vcvars` makes sense as an argument to `with tools.vcvars(..., post=self._append_vcvars)` and keep `environment_append` unmodified.